examples/jsm/renderers/webgpu/WebGPUNodeBuilder.js

```diff
@@ -0,0 +1,164 @@
+import WebGPUUniformsGroup from './WebGPUUniformsGroup.js';
+import { FloatUniform, Vector3Uniform } from './WebGPUUniform.js';
+
+import NodeSlot from '../nodes/core/NodeSlot.js';
+import NodeBuilder from '../nodes/core/NodeBuilder.js';
+
+class WebGPUNodeUniformsGroup extends WebGPUUniformsGroup {
+	
+	constructor( shaderStage ) {
+
+		super( 'nodeUniforms' );
+		
+		let shaderStageVisibility;
+		
+		if ( shaderStage === 'vertex' ) shaderStageVisibility = GPUShaderStage.VERTEX;
+		else if ( shaderStage === 'fragment' ) shaderStageVisibility = GPUShaderStage.FRAGMENT;
+			
+		this.setVisibility( shaderStageVisibility );
+		
+		this.setOnBeforeUpdate( this._onBeforeUpdate );
+		
+	}
+	
+	_onBeforeUpdate( object, camera ) {
+		
+		const material = object.material;
+		
+		console.log( 1 );
+		
+	}
+	
+}
+
+class WebGPUNodeBuilder extends NodeBuilder {
+
+	constructor( material, renderer ) {
+
+		super( material, renderer );
+		
+		this.uniformsGroup = {};
+		
+		this._parseMaterial();
+		
+	}
+	
+	_parseMaterial() {
+		
+		const material = this.material;
+		
+		if ( material.isMeshBasicMaterial || material.isPointsMaterial ) {
+			
+			if ( material.colorNode !== undefined ) {
+				
+				this.addSlot( 'fragment', new NodeSlot( material.colorNode, 'COLOR', 'vec3' ) );
+				
+			}
+			
+			if ( material.opacityNode !== undefined ) {
+				
+				this.addSlot( 'fragment', new NodeSlot( material.opacityNode, 'OPACITY', 'float' ) );
+				
+			}
+			
+		}
+		
+	}
+	
+	getUniformNSName( nodeUniform ) {
+		
+		return `nodeUniforms.${nodeUniform.name}`;
+		
+	}
+	
+	getBindings() {
+		
+		const bindings = [];
+		
+		const uniformsVertexGroup = this.uniformsGroup[ 'vertex' ];
+		const uniformsFragmentGroup = this.uniformsGroup[ 'fragment' ];
+		
+		if ( uniformsVertexGroup ) bindings.push( uniformsVertexGroup );
+		if ( uniformsFragmentGroup ) bindings.push( uniformsFragmentGroup );
+		
+		return bindings;
+		
+	}
+	
+	createUniformFromNode( node, shaderStage, type ) {
+		
+		const uniformNode = super.createUniformFromNode( node, shaderStage, type ) 
+		const nodeData = this.createDataFromNode( node, shaderStage );
```

Comment:
This call irritated me a bit. You call `createDataFromNode()` here although this method was already called inside of `createUniformFromNode()`. I understand that you only call it to _get_ the (already existing) node data. 

This code will be more clear when it is align to `WebGPUProperties` or `WebGLProperties`. Meaning to transform it to a getter.  The implementation is already similar to:

https://github.com/mrdoob/three.js/blob/b1390eb11be57c66b1239b822039198621ecb500/src/renderers/webgl/WebGLProperties.js#L5-L18
Meaning the method creates a cached object if necessary. But this creation is not something the caller has not know about. I suggest to use `getUniformNodeData()`. The code will look like so then:
```js
const uniformNode = super.getUniformFromNode( node, shaderStage, type ) 
const nodeData = this.getUniformNodeData( node, shaderStage );
```
In my mind `create*()` methods are usually factory methods that always return a new instance when called. This would is currently not true for both uniform methods.

Reply:
This was get before `getNodeData` I will keep and I have that change the logic of  `createUniformFromNode`... This code block changed a lot from yesterday to today because of the implementation of texture2d uniform...